fili-core/src/test/groovy/com/yahoo/bard/webservice/web/endpoints/JobsServletReactiveChainforResultsEndpointSpec.groovy

```diff
@@ -152,73 +181,119 @@ class JobsServletReactiveChainforResultsEndpointSpec extends Specification {
+        and:
+        PreResponse preResponse = PreResponseTestingUtils.buildPreResponse("2016-04-24")
 
-        and: "We miss the notification that the preResponse is stored in the PreResponseStore"
+        when: "We miss the notification that the preResponse is stored in the PreResponseStore"
         broadcastChannel.publish("ticket4")
 
         and: "We start the async chain"
-        mockJobServlet.getPreResponseObservable("ticket4", apiRequest1)
-        Thread.sleep(1000)
+        mockJobServlet.getResults("ticket4", apiRequest1.asyncAfter).subscribe()
 
-        then: "then we go to the PreResponseStore exactly once to get the ticket"
-        1 * mockPreResponseStore.get(_) >> Observable.just(Mock(PreResponse))
+        then: "Then we go to the PreResponseStore exactly once to get the ticket"
+        1 * mockPreResponseStore.get(_) >> Observable.empty()
     }
 
-    def "If the PreResponse is available in the PreResponsestore and the notification from broadcastChannel is not received within async timeout, we go to the PreResponseStore exactly once"() {
-        setup:
-        JobsApiRequest apiRequest1 = new JobsApiRequest(null, "2", "", "", null, uriInfo, jobPayloadBuilder, apiJobStore, mockPreResponseStore, broadcastChannel)
+    def "When multiple results-are-ready notifications are sent, we only receive the results we care about"() {
+        given: "A proper PreResponseStore that allows us to know when values appear in the store"
+        CountDownLatch getTicket1Latch = new CountDownLatch(1)
+        preResponseStore.addGetLatch("ticket1", getTicket1Latch)
+        CountDownLatch saveTicket1Latch = new CountDownLatch(1)
+        preResponseStore.addSaveLatch("ticket1", saveTicket1Latch)
+
+        and:
+        JobsApiRequest apiRequest = new JobsApiRequest(
+                null,
+                "never",
+                "",
+                "",
+                null,
+                uriInfo,
+                jobPayloadBuilder,
+                apiJobStore
+        )
 
-        when: "PreResponse is available in the PreResponsestore"
-        mockPreResponseStore.save("ticket4", PreResponseTestingUtils.buildPreResponse("2016-04-24")).subscribe()
+        and:
+        TestSubscriber resultsSubscriber = new TestSubscriber()
 
-        and: "We start the async chain"
-        mockJobServlet.getPreResponseObservable("ticket4", apiRequest1)
-        //The delay is to ensure that we get the notification after async timeout
-        Thread.sleep(1000)
+        and:
+        jobsServlet.getResults("ticket1", apiRequest.asyncAfter).subscribe(resultsSubscriber)
 
-        and: "We receive the notification after async timeout"
-        broadcastChannel.publish("ticket4")
+        and: "We wait for the first attempt to get results from the store to come up empty before we add fake results"
+        getTicket1Latch.await(30, TimeUnit.SECONDS)
```

Comment:
So that the test doesn't hang indefinitely in case something goes wrong and we never actually countdown the latches. I've moved that timeout into a timeout annotation on the test.